### PR TITLE
Little funny mapper

### PR DESCRIPTION
**Pull request by EgorkaZ**

### CODE REVIEW

File: src/alphabet_map.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+use std::iter::FromIterator;
+
+use rustc_hash::FxHashMap;
```
Comment (EgorkaZ) [inline src/alphabet_map.rs:3]: Exactly casual `HashMap`, but has a faster hash function. Should be good for integral type like `char`

File: src/lib.rs [hunk 2]
```diff
@@ -1,38 +1,48 @@
...
-        to_index(ch);
-    });
+fn check_alphabet(input: &str) -> Result<&str, char> {
+    input.chars()
+        .find(|ch| !ALPHABET_MAP.contains_char(*ch))
+        .map_or(Ok(input), Err)
+}
+
+fn panic_on_bad_char<Out>(ch: char) -> Out {
+    panic!("Character {} is not found in alphabet: [{}]", ch, ALPHABET)
```
Comment (BobIsOnFire) [inline src/lib.rs:29]: Could you please return char codes into panic message? I could see a confusion if it says that char 'А' (cyrillic) is not found in alphabet [ABC] (latin)

File: src/lib.rs [hunk 3]
```diff
@@ -44,23 +54,17 @@ impl CaesarSolver {
         CaesarSolver { letter_mapper }
     }
 
-    pub fn encrypt(&self, input: String) -> String {
```
Comment (BobIsOnFire) [inline src/lib.rs:47]: We should have interfaces for both `String` and `Iterator`. I love the possibility to read from char stream or something like that, but most of the cases are actually about transforming strings and they cause additional function calls outside of the interface.
How about `encrypt` and `encrypt_stream`?

File: src/lib.rs [hunk 4]
```diff
@@ -44,23 +54,17 @@ impl CaesarSolver {
...
-        let input = input.to_uppercase();
-        check_alphabet(&input);
```
Comment (BobIsOnFire) [inline src/lib.rs:49]: Is it ever checked that input is a part of alphabet? Apart from `by_idx` call which does not fail gracefully

**Review by BobIsOnFire — CHANGES_REQUESTED** [review]
<3

**Review by BobIsOnFire — APPROVED** [review]

**Pull request merged**